controllers/stardoginstance_controller.go

```diff
@@ -7,36 +7,293 @@ package controllers
+		controllerutil.RemoveFinalizer(stardogInstance, stardogRolesFinalizer)
+		err := r.Update(rc.ctx, stardogInstance)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) finalize(rc *ReconciliationContext, finalizer string) error {
+	username, password, err := r.getCredentials(rc.ctx, rc.cfg.Spec.AdminCredentials)
+	if err != nil {
+		return err
+	}
+
+	stardog := rest.NewWithBaseURIAndAuthentication(rc.cfg.Spec.ServerUrl, username, password)
+	if finalizer == stardogUsersFinalizer {
+		users, err := stardog.ListUsers(rc.ctx)
+		if err != nil {
+			r.Log.Error(err, "cannot list Stardog users", "namespace", rc.cfg.Namespace)
+			return err
+		}
+		if len(*users.Users) > 0 {
+			return fmt.Errorf("cannot delete StardogInstance, found %d users", len(*users.Users))
+		}
+	}
+	if finalizer == stardogRolesFinalizer {
+		roles, err := stardog.ListRoles(rc.ctx)
+		if err != nil {
+			r.Log.Error(err, "cannot list Stardog users", "namespace", rc.cfg.Namespace)
+			return err
+		}
+		if len(*roles.Roles) > 0 {
+			return fmt.Errorf("cannot delete StardogInstance, found %d roles", len(*roles.Roles))
+		}
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) validateSpecification(spec StardogInstanceSpec) error {
+	r.Log.Info("validating specification of StardogInstance CRD")
+	if len(spec.ServerUrl) == 0 {
+		return fmt.Errorf(".spec.ServerUrl is required")
+	}
+	if len(spec.AdminCredentials.SecretRef) == 0 {
+		return fmt.Errorf(".spec.AdminCredentials.SecretRef is required")
+	}
+	_, err := url.ParseRequestURI(spec.ServerUrl)
+	if err != nil {
+		return fmt.Errorf(".spec.ServerUrl is not a valid URL")
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) validateConnection(rc *ReconciliationContext) error {
+	r.Log.Info("validating Connection to Stardog")
+	if len(rc.cfg.Spec.AdminCredentials.Namespace) == 0 {
+		rc.cfg.Spec.AdminCredentials.Namespace = rc.cfg.Namespace
+	}
+	username, password, err := r.getCredentials(rc.ctx, rc.cfg.Spec.AdminCredentials)
+	if err != nil {
+		return err
+	}
+	stardog := rest.NewWithBaseURIAndAuthentication(rc.cfg.Spec.ServerUrl, username, password)
+	_, err = stardog.IsEnabled(rc.ctx, username)
+	if err != nil {
+		r.Log.Error(err, "stardog is not reachable")
+	}
+	return nil
+}
 
-	return ctrl.Result{}, nil
+func (r *StardogInstanceReconciler) getCredentials(ctx context.Context, credentials StardogUserCredentialsSpec) (username, password string, err error) {
+	r.Log.Info("retrieving admin credentials from", "secret", credentials.SecretRef)
+	secret := &v1.Secret{}
+	err = r.Client.Get(ctx, types.NamespacedName{Namespace: credentials.Namespace, Name: "Secret"}, secret)
+	if err != nil {
+		r.Log.Error(err, "cannot retrieve admin credentials from", "Secret", credentials.SecretRef)
+		return "", "", err
+	}
+	user := secret.StringData["username"]
+	if len(username) == 0 {
+		return "", "", fmt.Errorf(".spec.AdminCredentials.SecretRef.username is required")
+	}
```

Comment:
Same applies for password
```suggestion
	if user == "" {
		return "", "", fmt.Errorf(".data.username in the Secret %s/%s is required", credentials.Namespace, credentials.Name)
	}
```